File: 448_Find_All_Numbers_Disappeared_in_an_Array.java
/*
https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
Given an array of integers where 1 ≤ a[i] ≤ n (n = size of array), some elements appear twice and others appear once.
Find all the elements of [1, n] inclusive that do not appear in this array.


Time - O(n)
Space - O(1)
*/

public class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        for (int num : nums) {
            int i = Math.abs(num) - 1;      // num may have already have sign flipped so take absolute value
            if (nums[i] > 0)
                nums[i] = -nums[i];
        }

        List<Integer> missing = new LinkedList<Integer>();
        for (int i = 0; i < nums.length; ++i)
            if (nums[i] > 0)
                missing.add(i + 1);

        return missing;
    }
}

File: python_1_to_1000/448_Find_All_Numbers_Disappeared_in_an_Array.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/
# Given an array of integers where 1 ≤ a[i] ≤ n (n = size of array), some elements appear twice and others appear once.
# Find all the elements of [1, n] inclusive that do not appear in this array.

# Iterate over the list. For every number seen, set the number with index of num - 1 to be negative.
# Iterate again over the list, finding all indices that are still positive and so have not been seen.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def findDisappearedNumbers(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        for num in nums:
            num = abs(num)                          # may be negative so take absolute value
            nums[num - 1] = -abs(nums[num - 1])     # set nums[num - 1] to indicate num - 1 seen

        return [i + 1 for i, num in enumerate(nums) if num > 0]
